app/vmctl/mimir/mimir.go

```diff
@@ -0,0 +1,240 @@
+	min, max   int64
+	label      string
+	labelValue string
+}
+
+func (f filter) inRange(min, max int64) bool {
+	fmin, fmax := f.min, f.max
+	if min == 0 {
+		fmin = min
+	}
+	if fmax == 0 {
+		fmax = max
+	}
+	return min <= fmax && fmin <= max
+}
+
+// NewClient creates and validates new Client
+// with given Config
+func NewClient(cfg Config) (*Client, error) {
+	if cfg.Path == "" {
+		return nil, fmt.Errorf("path cannot be empty")
+	}
+
+	if cfg.TenantID != "" {
+		cfg.Path = fmt.Sprintf("%s/%s", cfg.Path, cfg.TenantID)
+	}
+
+	var c Client
+	rfs, err := NewRemoteFS(cfg)
+	if err != nil {
+		return nil, fmt.Errorf("cannot parse `-src`=%q: %w", cfg.Path, err)
+	}
+
+	c.RemoteFS = rfs
+	timeMin, err := utils.ParseTime(cfg.Filter.TimeMin)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse min time in filter: %s", err)
+	}
+	timeMax, err := utils.ParseTime(cfg.Filter.TimeMax)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse max time in filter: %s", err)
+	}
+	c.filter = filter{
+		min:        timeMin.UnixMilli(),
+		max:        timeMax.UnixMilli(),
+		label:      cfg.Filter.Label,
+		labelValue: cfg.Filter.LabelValue,
+	}
+	return &c, nil
+}
+
+// Explore fetches all available blocks from a remote storage or local filesystem
+// and collects the Meta() data from each block.
+// Explore does initial filtering by time-range
+// for snapshot blocks but does not take into account
+// label filters.
+func (c *Client) Explore() ([]tsdb.BlockReader, error) {
+
+	s := &utils.Stats{
+		Filtered: c.filter.min != 0 || c.filter.max != 0 || c.filter.label != "",
+	}
+
+	log.Printf("Fetching blocks from remote storage")
+
+	indexFile, err := c.fetchIndexFile()
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch index file: %s", err)
+	}
+
+	var blocksToImport []tsdb.BlockReader
+	for _, block := range indexFile.Blocks {
+		if !c.filter.inRange(block.MinTime, block.MaxTime) {
+			// Skipping block outside of time range
+			continue
+		}
+
+		if block.ID.String() == "" {
+			continue
+		}
+
+		lazyBlockReader, err := New(block, c.RemoteFS)
+		if err != nil {
+			log.Printf("failed to create lazy block reader: %s", err)
+			continue
+		}
+		blocksToImport = append(blocksToImport, lazyBlockReader)
+	}
+
+	s.Blocks = len(blocksToImport)
+	return blocksToImport, nil
+}
+
+// Read reads the given BlockReader according to configured
+// time and label filters.
+func (c *Client) Read(block tsdb.BlockReader) (storage.SeriesSet, error) {
+	meta := block.Meta()
+	if meta.ULID.String() == "" {
+		return nil, fmt.Errorf("block without id")
+	}
+
+	minTime, maxTime := meta.MinTime, meta.MaxTime
+	if c.filter.min != 0 {
+		minTime = c.filter.min
+	}
+	if c.filter.max != 0 {
+		maxTime = c.filter.max
+	}
+	q, err := tsdb.NewBlockQuerier(block, minTime, maxTime)
+	if err != nil {
+		return nil, err
+	}
+	ss := q.Select(context.Background(), false, nil, labels.MustNewMatcher(labels.MatchRegexp, c.filter.label, c.filter.labelValue))
```

Comment:
It seems like `q.Select` can block client for quite long period of time so it would be great to pass `ctx` to the client so that cancelling an import would result in `q.Select` cancellation.